include/ginkgo/core/matrix/sparsity.hpp

```diff
@@ -0,0 +1,218 @@
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_MATRIX_SPARSITY_HPP_
+#define GKO_CORE_MATRIX_SPARSITY_HPP_
+
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+
+
+namespace gko {
+namespace matrix {
+
+
+/**
+ * SPARSITY is a matrix format which stores only the sparsity pattern of a
+ * sparse matrix by compressing each row of the matrix (compressed sparse row
+ * format).
+ *
+ * The values of the nonzero elements are not stored because they are assumed to
+ * be one by default. A row pointer array which stores the starting index of
+ * each row. An additional column index array is used to identify the column
+ * where a nonzero is present.
+ *
+ * @tparam ValueType  precision of vectors in apply
+ * @tparam IndexType  precision of matrix indexes
+ *
+ * @ingroup sparsity
+ * @ingroup mat_formats
+ * @ingroup LinOp
+ */
+template <typename ValueType = default_precision, typename IndexType = int32>
+class Sparsity : public EnableLinOp<Sparsity<ValueType, IndexType>>,
+                 public EnableCreateMethod<Sparsity<ValueType, IndexType>>,
+                 public ReadableFromMatrixData<ValueType, IndexType>,
+                 public WritableToMatrixData<ValueType, IndexType>,
+                 public Transposable {
+    friend class EnableCreateMethod<Sparsity>;
+    friend class EnablePolymorphicObject<Sparsity, LinOp>;
+
+public:
+    using EnableLinOp<Sparsity>::convert_to;
+    using EnableLinOp<Sparsity>::move_to;
+
+    using index_type = IndexType;
+
+    using mat_data = matrix_data<ValueType, IndexType>;
+
+    void read(const mat_data &data) override;
+
+    void write(mat_data &data) const override;
+
+    std::unique_ptr<LinOp> transpose() const override;
+
+    std::unique_ptr<LinOp> conj_transpose() const override;
+
+    /**
+     * Transforms the sparsity matrix to an adjacency matrix.
+     *
```

Comment:
Can an adjacency matrix be built by non-square matrix?
If it is not, it would be great to add the description here and add the assert in this function.